### PR TITLE
[deprecated]: [DO NOT MERGE] Install OpenEBS on a Director Onprem Cluster

### PR DESCRIPTION
**Pull request by uditgaurav**
Signed-off-by: Udit Gaurav <uditgaurav@gmail.com>
issue - #198 
**_This PR is for:_**

- Automating the openebs installation process in DOP cluster

**_Folder added:_**
- /oep/litmus/director/openebs-self-install

**_Details:_**

**_Steps involved in openebs installation in the self dop clusters:_**

- **_Labeling the Nodes_**: Nodes are labeled for the control and data components.
  - `GET` request to the nodes on success -
  - `POST` request for the change in labels according to the condition - On which node you want to place the data components and on which node you want to place the control plan components.
- **_Generating the request for openebs installation:_**
  - A request is generated for openebs installation
  - `POST` request for openebs installation:
  - The `GET`  Respond from the previous step (in JSON format) is sent as Request to the for the installation. 

**_Output of the ansible job:_**
- [X] [logs.txt](https://github.com/mayadata-io/oep/files/4217565/logs.txt)

### CODE REVIEW

**Comment by AmitKumarDas** [comment]
@uditgaurav can you please change the signing email to mayadata account

File: litmus/director/openebs-self-install/run_litmus_test.yml [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+      name: litmus
+      labels:
+        app: openebs-self-install-check-litmus
+    spec:
+      serviceAccountName: litmus
+      restartPolicy: Never
+      volumes:
+      - name: secret-volume
+        secret:
+          secretName: director-user-pass
```
Comment (AmitKumarDas) [inline litmus/director/openebs-self-install/run_litmus_test.yml]: can we add comment on why this secret is required?

File: litmus/director/openebs-self-install/run_litmus_test.yml [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+      containers:
+      - name: ansibletest
```
Comment (AmitKumarDas) [inline litmus/director/openebs-self-install/run_litmus_test.yml]: Can we add some yaml comments on what dop-validator image does?

File: litmus/director/openebs-self-install/run_litmus_test.yml [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+        image: mayadataio/dop-validator:ci
```
Comment (AmitKumarDas) [inline litmus/director/openebs-self-install/run_litmus_test.yml]: why dont we use a proper release tag?

File: litmus/director/openebs-self-install/run_litmus_test.yml [hunk 4]
```diff
@@ -0,0 +1,95 @@
...
+        imagePullPolicy: Always
+        volumeMounts:
+        - name: secret-volume
+          readOnly: true
+          mountPath: "/etc/secret-volume"
+        env:
+          
+          ## Take url from configmap config
```
Comment (AmitKumarDas) [inline litmus/director/openebs-self-install/run_litmus_test.yml]: Do we have this mentioned in somewhere in the readme of this test case.

File: litmus/director/openebs-self-install/run_litmus_test.yml (hunk 2 above)
Comment (AmitKumarDas) [reply litmus/director/openebs-self-install/run_litmus_test.yml]: Do we have this mentioned in somewhere in the readme of this test case.

File: litmus/director/openebs-self-install/run_litmus_test.yml [hunk 5]
```diff
@@ -0,0 +1,95 @@
...
+
+          ## Takes group_id from configmap groupid
+          - name: GROUP_ID
+            valueFrom:
+              configMapKeyRef:
+                name: groupid
+                key: group_id
+
+          ## It should be 1ot1 Mandatory
+          - name: TEMPLATE_ID
```
Comment (AmitKumarDas) [inline litmus/director/openebs-self-install/run_litmus_test.yml]: What is the significance of this? Can we add yaml comment as well as mention / describe this in the README.

File: litmus/director/openebs-self-install/run_litmus_test.yml [hunk 6]
```diff
@@ -0,0 +1,95 @@
...
+            value: '1ot1'
+
+          ## Namespace where openebs is installed
+          ## By default in basic mode it is openebs
+          - name: NAMESPACE
```
Comment (AmitKumarDas) [inline litmus/director/openebs-self-install/run_litmus_test.yml]: **[MUST]** IMO this is most important. We should install both in basic mode & non basic mode.
cc @ajeshbaby

Comment (AmitKumarDas) [reply litmus/director/openebs-self-install/run_litmus_test.yml]: Do we have these modes (e.g. basic, etc) mentioned /described in somewhere in the readme of this test case.

File: litmus/director/openebs-self-install/run_litmus_test.yml [hunk 7]
```diff
@@ -0,0 +1,95 @@
...
+            value: ''
+
+          ## Enter the default directory - It can be /var/openebs
+          - name: DEFAULT_DIRECTORY
```
Comment (AmitKumarDas) [inline litmus/director/openebs-self-install/run_litmus_test.yml]: Do we have these mentioned /described in somewhere in the readme of this test case.
We need to explain what default directory means? How is this related to jiva or cstor or local pv? Is this related to storage engines at all?

File: litmus/director/openebs-self-install/run_litmus_test.yml [hunk 8]
```diff
@@ -0,0 +1,95 @@
...
+            value: ''
+
+          ##Enter docker registry
```
Comment (AmitKumarDas) [inline litmus/director/openebs-self-install/run_litmus_test.yml]: Let us explain these entries & mention what values has been used actually in the README.

File: litmus/director/openebs-self-install/run_litmus_test.yml [hunk 9]
```diff
@@ -0,0 +1,95 @@
...
+          - name: MEMORY_RESOURCE_LIMIT
+            value: ''
+          
+          ## It will have values basic/advance
+          ## Note: For basic installation only template value slould be provided
+          ## For adding any additional value - change the installation mode to advance
+          - name: INSTALLATION_MODE
+            value: 'basic'
+
+          - name: ANSIBLE_STDOUT_CALLBACK
```
Comment (AmitKumarDas) [inline litmus/director/openebs-self-install/run_litmus_test.yml]: Do we have these mentioned /described in somewhere in the readme of this test case.

File: litmus/director/openebs-self-install/run_litmus_test.yml (hunk 2 above)
Comment (uditgaurav) [reply litmus/director/openebs-self-install/run_litmus_test.yml]: The `secrets` and `configmaps` have been created in the early stages so the readmes are there only.

File: litmus/director/openebs-self-install/run_litmus_test.yml [hunk 10]
```diff
@@ -0,0 +1,95 @@
...
+            value: 'default'  
+
+        command: ["/bin/bash"]
+        args: ["-c", "ansible-playbook ./litmus/director/openebs-self-install/test.yml -i /etc/ansible/hosts -v; exit 0"]
```
Comment (AmitKumarDas) [inline litmus/director/openebs-self-install/run_litmus_test.yml]: is it possible to write the entire command in a list format?
e.g.
```yaml
args:
- "c"
- "ansible-playbook"
- "./litmus/director/openebs-self-install/test.yml -i /etc/ansible/hosts -v; exit 0"
```

File: litmus/director/openebs-self-install/run_litmus_test.yml (hunk 4 above)
Comment (uditgaurav) [reply litmus/director/openebs-self-install/run_litmus_test.yml]: Yes in the readme of the test case I'll be mentioning this!

File: litmus/director/openebs-self-install/test.yml [hunk 11]
```diff
@@ -0,0 +1,187 @@
+---
+- hosts: localhost
```
Comment (AmitKumarDas) [inline litmus/director/openebs-self-install/test.yml]: Let us put in yaml comments description of each & every field.

File: litmus/director/openebs-self-install/run_litmus_test.yml (hunk 5 above)
Comment (uditgaurav) [reply litmus/director/openebs-self-install/run_litmus_test.yml]: I'll be adding in the readme about every `envs`, `utils` which it uses, about `secrets` and `configmaps`

File: litmus/director/openebs-self-install/run_litmus_test.yml (hunk 6 above)
Comment (uditgaurav) [reply litmus/director/openebs-self-install/run_litmus_test.yml]: I have tested it in both `basic` and `nonbasic` mode and give some default values (i.e basic mode)

File: litmus/director/openebs-self-install/run_litmus_test.yml (hunk 7 above)
Comment (uditgaurav) [reply litmus/director/openebs-self-install/run_litmus_test.yml]: Yes, @AmitKumarDas  so should we have one `README` here and one in the `oep-e2e-gcp` repo or we can maintain only one.

File: litmus/director/openebs-self-install/test.yml [hunk 12]
```diff
@@ -0,0 +1,187 @@
...
+    ## Fetch the projectid of the project where the cluster is running
+    - name: Fetch the project id of the project where the cluster is running
+      uri:
+        url: "{{ director_url }}/v3/groups/{{ group_id }}/project"
+        method: GET
+        url_username: "{{ username.stdout }}"
+        url_password: "{{ password.stdout }}"
+        force_basic_auth: yes
+        return_content: yes
+        status_code: 200
```
Comment (AmitKumarDas) [inline litmus/director/openebs-self-install/test.yml]: why is status_code required ?

File: litmus/director/openebs-self-install/test.yml [hunk 13]
```diff
@@ -0,0 +1,187 @@
...
+        executable: /bin/bash
+      register: maya_api
+      failed_when: "{{ maya_api.stdout }} != 0"
+
+    ## Fetch the projectid of the project where the cluster is running
+    - name: Fetch the project id of the project where the cluster is running
+      uri:
+        url: "{{ director_url }}/v3/groups/{{ group_id }}/project"
+        method: GET
+        url_username: "{{ username.stdout }}"
```
Comment (AmitKumarDas) [inline litmus/director/openebs-self-install/test.yml]: [Question]: What all possible auth modes are available?

File: litmus/director/openebs-self-install/run_litmus_test.yml (hunk 10 above)
Comment (uditgaurav) [reply litmus/director/openebs-self-install/run_litmus_test.yml]: Yes mabe like this 
```
args:
- "-c"
- "ansible-playbook ./litmus/director/openebs-self-install/test.yml -i /etc/ansible/hosts -v"
- "exit 0"

Is that a better approach ?

File: litmus/director/openebs-self-install/test.yml (hunk 13 above)
Comment (AmitKumarDas) [reply litmus/director/openebs-self-install/test.yml]: [Question]: Can we use some secrets to authenticate?

File: litmus/director/openebs-self-install/test.yml (hunk 12 above)
Comment (uditgaurav) [reply litmus/director/openebs-self-install/test.yml]: So whatever response we get as a result of  `GET` or `POST` request it comes with status code which is basically predefined and we catch them using `status code`  field in ansible `uri` module. status code `200` is default status code if I remove this then also it will not make any difference.

File: litmus/director/openebs-self-install/test.yml (hunk 13 above)
Comment (uditgaurav) [reply litmus/director/openebs-self-install/test.yml]: Right now we are using `basic auth` only.

File: litmus/director/openebs-self-install/run_litmus_test.yml (hunk 3 above)
Comment (uditgaurav) [reply litmus/director/openebs-self-install/run_litmus_test.yml]: We have used `ci` tag in all other tests. 
should we use proper release tag here `i.e 1.8.0`? @harshshekhar15

File: litmus/director/openebs-self-install/test.yml [hunk 14]
```diff
@@ -0,0 +1,161 @@
...
+      shell: kubectl get nodes | wc -l
+      args:
+        executable: /bin/bash
+      register: nodes
+
+    - fail:
+        msg: Minimum two nodes is required to install openebs in this cluster
+      when: "{{ nodes.stdout }} < 2"
+
+    - block:
```
[the remark this answers is not in the document]
Comment (harshshekhar15) [reply litmus/director/openebs-self-install/test.yml]: @uditgaurav please put all the tasks within `block` is what I meant.

File: litmus/director/openebs-self-install/test.yml (hunk 13 above)
Comment (uditgaurav) [reply litmus/director/openebs-self-install/test.yml]: @AmitKumarDas  we are using secrets only. If you will go to `run_litmus_test.yml` line 19 a secret called `director-user-pass` is used which has been created in the early stages. We are only using that here.

```
    - name: Get username
      shell: cat /etc/secret-volume/username
      register: username
```
we are taking the username from secret volume

File: litmus/director/openebs-self-install/run_litmus_test.yml (hunk 3 above)
Comment (harshshekhar15) [reply litmus/director/openebs-self-install/run_litmus_test.yml]: @AmitKumarDas this image gets contains all the test cases that are present in this repo. It gets updated with the latest test cases after every PR merge in this repo. This image continuously gets updated on every merge so I am not so sure if we should use a release tag here.

File: litmus/director/openebs-self-install/run_litmus_test.yml (hunk 6 above)
Comment (uditgaurav) [reply litmus/director/openebs-self-install/run_litmus_test.yml]: we have two modes `basic` and `advance` mode.

File: litmus/director/openebs-self-install/run_litmus_test.yml (hunk 3 above)
Comment (uditgaurav) [reply litmus/director/openebs-self-install/run_litmus_test.yml]: okay done @harshshekhar15

**Comment by uditgaurav** [comment]
closing the PR due to some Github issue in this PR
